plugins/macaron/macaron.go

```diff
@@ -0,0 +1,79 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package macaron
+
+import (
+	"fmt"
+	"net/http"
+
+	"gopkg.in/macaron.v1"
+
+	"go.opentelemetry.io/contrib/internal/trace"
+	otelglobal "go.opentelemetry.io/otel/api/global"
+	otelpropagation "go.opentelemetry.io/otel/api/propagation"
+	oteltrace "go.opentelemetry.io/otel/api/trace"
+)
+
+const (
+	tracerKey  = "otel-go-contrib-tracer"
+	tracerName = "go.opentelemetry.io/contrib/plugins/macaron"
+)
+
+// Middleware returns stuff
+func Middleware(service string, opts ...Option) macaron.Handler {
+	cfg := Config{}
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+	if cfg.Tracer == nil {
+		cfg.Tracer = otelglobal.Tracer(tracerName)
+	}
+	if cfg.Propagators == nil {
+		cfg.Propagators = otelglobal.Propagators()
+	}
+	return func(res http.ResponseWriter, req *http.Request, c *macaron.Context) {
+		savedCtx := c.Req.Request.Context()
+		defer func() {
+			c.Req.Request = c.Req.Request.WithContext(savedCtx)
+		}()
+
+		ctx := otelpropagation.ExtractHTTP(savedCtx, cfg.Propagators, c.Req.Header)
+		opts := []oteltrace.StartOption{
+			oteltrace.WithAttributes(trace.NetAttributesFromHTTPRequest("tcp", c.Req.Request)...),
+			oteltrace.WithAttributes(trace.EndUserAttributesFromHTTPRequest(c.Req.Request)...),
+			oteltrace.WithAttributes(trace.HTTPServerAttributesFromHTTPRequest(service, c.Req.RequestURI, c.Req.Request)...),
+			oteltrace.WithSpanKind(oteltrace.SpanKindServer),
+		}
+		// TODO: span name should be router template not the actual request path, eg /user/:id vs /user/123
+		spanName := c.Req.RequestURI
```

Comment:
There indeed seems to be no way to get the matched route. Boo. I think we should give a different name for the span then - the request URI is going to have a too high cardinality because of the parameters in the uri.

I see two ways out here:

1. convince macaron authors to provide the functionality for getting the matched route
    - that's a long way to resolve the issue - one thing is providing functionality, another thing is for the project to bump the macaron dependency to the new version
1. provide span name generator function as an option to `Middleware`, so `Middleware` could use it to get the span name.

The last option could be something like:

```go
// in config.go
type NameGenerator func(c *macaron.Context) string

func WithNameGenerator(g NameGenerator) Option {…}
```

So the example server could call `Middleware` like so:

```go
nameGenerator := func(c *macaron.Context) string {
	uri := c.Req.RequestURI
	if strings.HasPrefix(c.Req.RequestURI, "/users/") {
		return "/users/:id"
	}
	return "unknown route"
}
m.Use(macarontrace.Middleware("my-server", macarontrace.WithNameGenerator(nameGenerator)))
```

It's lame, but I guess you pay for what you use.

Reply:
https://github.com/open-telemetry/opentelemetry-go-contrib/issues/37